tests/mqtt_mock_server_handler.h

```diff
@@ -407,3 +467,152 @@ static void s_mqtt_mock_server_wait_for_pubacks(struct aws_channel_handler *hand
     aws_condition_variable_wait_pred(&testing_handler->cvar, &testing_handler->lock, s_is_pubacks_complete, &waiter);
     aws_mutex_unlock(&testing_handler->lock);
 }
+
+/* this is only safe to call when not attached to a channel. */
+struct aws_array_list *mqtt_mock_server_get_received_messages(struct aws_channel_handler *handler) {
+    struct mqtt_mock_server_handler *testing_handler = handler->impl;
+
+    return &testing_handler->received_messages;
+}
+
+/* dynamic append from to to, the cursor of the buffer we just appended stored in result */
+static int s_move_cursor_to_buf_dynamic(
+    struct aws_byte_buf *to,
+    struct aws_byte_cursor *from,
+    struct aws_byte_cursor *result) {
+
+    size_t pre_len = to->len;
+    (void)pre_len;
+    /* move the cursor to buffer */
+    ASSERT_SUCCESS(aws_byte_buf_append_dynamic(to, from));
+    *result = aws_byte_cursor_from_buf(to);
+    /* move the result cursor to point at the place we just moved to */
+    aws_byte_cursor_advance(result, pre_len);
+
+    return AWS_OP_SUCCESS;
+}
+
+size_t mqtt_decoded_packets_count(struct aws_channel_handler *handler) {
+    struct mqtt_mock_server_handler *testing_handler = handler->impl;
+    return aws_array_list_length(&testing_handler->packets);
+}
+
+struct mqtt_decoded_packet *mqtt_get_decoded_packet(struct aws_channel_handler *handler, size_t i) {
+    struct mqtt_mock_server_handler *testing_handler = handler->impl;
+    AWS_FATAL_ASSERT(mqtt_decoded_packets_count(handler) > i);
+    struct mqtt_decoded_packet *packet = NULL;
+    aws_array_list_get_at_ptr(&testing_handler->packets, (void **)&packet, i);
+    return packet;
+}
+
+struct mqtt_decoded_packet *mqtt_get_latest_decoded_packet(struct aws_channel_handler *handler) {
+    size_t packet_count = mqtt_decoded_packets_count(handler);
+    AWS_FATAL_ASSERT(packet_count > 0);
+    return mqtt_get_decoded_packet(handler, packet_count - 1);
+}
+
+int mqtt_mock_server_decoder_packets(struct aws_channel_handler *handler) {
+    struct mqtt_mock_server_handler *testing_handler = handler->impl;
+
+    struct aws_array_list received_messages = testing_handler->received_messages;
+    size_t length = aws_array_list_length(&received_messages);
+    for (size_t i = testing_handler->decoded_index; i < length; i++) {
+        struct aws_byte_buf received_message = {0};
+        ASSERT_SUCCESS(aws_array_list_get_at(&received_messages, &received_message, i));
+        struct aws_byte_cursor message_cur = aws_byte_cursor_from_buf(&received_message);
+
+        struct mqtt_decoded_packet packet;
+        ASSERT_SUCCESS(s_mqtt_decode_packet_init(&packet, testing_handler->handler.alloc));
+        packet.type = aws_mqtt_get_packet_type(message_cur.ptr);
+
+        struct aws_mqtt_packet_connect connect_packet;
```

Comment:
Probably be best to zero these.

Also, not necessary, but you could probably move these into the case statements if you added {} to them.

Reply:
Aha, that's what the complier complaining about! 🤯 Thank you!!!